loader/loader_windows.c

```diff
@@ -884,37 +884,106 @@ VkResult windows_read_sorted_physical_devices(struct loader_instance *inst, uint
                 } while ((vkres = icd_term->scanned_icd->EnumerateAdapterPhysicalDevices(
                               icd_term->instance, description.AdapterLuid, &count,
                               sorted_array[*sorted_devices_count].physical_devices)) == VK_INCOMPLETE);
-            }
 
-            if (vkres != VK_SUCCESS) {
-                loader_instance_heap_free(inst, sorted_array[*sorted_devices_count].physical_devices);
-                sorted_array[*sorted_devices_count].physical_devices = NULL;
-                if (vkres == VK_ERROR_OUT_OF_HOST_MEMORY) {
-                    res = VK_ERROR_OUT_OF_HOST_MEMORY;
-                    goto out;
-                } else {
-                    loader_log(inst, VULKAN_LOADER_WARN_BIT, 0, "Failed to convert DXGI adapter into Vulkan physical device");
-                    continue;
+                if (vkres != VK_SUCCESS) {
+                    loader_instance_heap_free(inst, sorted_array[*sorted_devices_count].physical_devices);
+                    sorted_array[*sorted_devices_count].physical_devices = NULL;
+                    if (vkres == VK_ERROR_OUT_OF_HOST_MEMORY) {
+                        res = VK_ERROR_OUT_OF_HOST_MEMORY;
+                        goto out;
+                    } else {
+                        loader_log(inst, VULKAN_LOADER_WARN_BIT, 0, "Failed to convert DXGI adapter into Vulkan physical device");
+                        continue;
+                    }
+                }
+
+                // Check if we have already added these physical devices - in case multiple adapters report the same LUID
+                bool already_has_phys_dev_per_icd = false;
+                for (uint32_t j = 0; j < *sorted_devices_count; j++) {
+                    if (count == sorted_array[j].device_count) {
+                        bool matches = true;
+                        for (uint32_t k = 0; k < sorted_array[j].device_count; k++) {
+                            if (sorted_array[j].physical_devices[k] != sorted_array[*sorted_devices_count].physical_devices[k]) {
+                                matches = false;
+
+                                break;
+                            }
+                        }
+                        if (matches) {
+                            already_has_phys_dev_per_icd = true;
+                        }
+                    }
+                }
+                if (!already_has_phys_dev_per_icd) {
+                    sorted_array[*sorted_devices_count].device_count = count;
```

Comment:
I would find it helpful if the name `sorted_array` was changed to `sorted_array_of_icds`. It seems to be an array of ICD info.

But then why is it indexed by a variable named `sorted_devices_count`? This asymmetry confuses me.

Reply:
Some of the variable naming predates my PR but I will take some time to rename things appropriately.